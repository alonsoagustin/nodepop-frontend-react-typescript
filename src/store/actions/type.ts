import { Advert } from "../../pages/types";

type Login = {
  type: "LOGIN";
};

type Logout = {
  type: "LOGOUT";
};

type AdvertsLoaded = {
  type: "ADVERTS_LOADED";
  payload: Advert[];
};

type AdvertsLoadedPending = {
  type: "ADVERTS_LOADED_PENDING";
};

type AdvertsLoadedFullfilled = {
  type: "ADVERTS_LOADED_FULFILLED";
  payload: { data: Advert[]; loaded: boolean };
};

type AdvertsLoadedRejected = {
  type: "ADVERTS_LOADED_REJECTED";
  payload: string;
};

type AdvertLoaded = {
  type: "ADVERT_LOADED";
};

type TagsLoaded = {
  type: "TAGS_LOADED";
};

type TagsLoadedPending = {
  type: "TAGS_LOADED_PENDING";
};

type TagsLoadedFullfilled = {
  type: "TAGS_LOADED_FULFILLED";
  payload: string[];
};

type TagsLoadedRejected = {
  type: "TAGS_LOADED_REJECTED";
  payload: string;
};

type FilterAdvertsByTag = {
  type: "FILTER_ADVERTS_BY_TAG";
  payload: string;
};

type FilterAdvertsByName = {
  type: "FILTER_ADVERTS_BY_NAME";
  payload: string;
};

type AdvertCreated = {
  type: "ADVERT_CREATED";
  payload: Advert;
};

type AdvertDeleted = {
  type: "ADVERT_DELETED";
  payload: number;
};

type Action =
  | Login
  | Logout
  | AdvertsLoaded
  | AdvertsLoadedPending
  | AdvertsLoadedFullfilled
  | AdvertsLoadedRejected
  | AdvertLoaded
  | FilterAdvertsByTag
  | FilterAdvertsByName
  | TagsLoaded
  | TagsLoadedPending
  | TagsLoadedFullfilled
  | TagsLoadedRejected
  | AdvertCreated
  | AdvertDeleted;

export default Action;
