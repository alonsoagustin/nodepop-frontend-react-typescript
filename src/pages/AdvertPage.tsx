import { useState, useEffect } from "react";
import { useParams, useNavigate, Navigate } from "react-router-dom";
import Button from "../components/shared/Button";
import Modal from "../components/shared/Modal";
import Spinner from "../components/shared/Spinner";
import { getAdvert, getUi } from "../store/selectors/selectors";
import { useAppDispatch, useAppSelector } from "../store/store";
import { AdvertLoaded, AdvertDeleted } from "../store/actions/creators";

const AdvertPage = () => {
  // Get the advert id from the URL
  const { advertId = "" } = useParams();

  // Get the advert from store using selector
  const advert = useAppSelector(getAdvert(advertId));

  // Get the loading state from store using selector
  const { error, loading } = useAppSelector(getUi);

  // Hook to dispatch actions
  const dispatch = useAppDispatch();

  // Hook to redirect to another page
  const navigate = useNavigate();

  // State to control the modal
  const [showModal, setShowModal] = useState(false);

  useEffect(() => {
    if (!advert) dispatch(AdvertLoaded(advertId));
  }, [dispatch, advert, advertId]);

  const handleConfirmDelete = async () => {
    setShowModal(false);
    await dispatch(AdvertDeleted(advertId, navigate));
  };

  return (
    <>
      {/* Modal */}
      <Modal
        title="Are you sure you want to delete this advert?"
        showModal={showModal}
        onClose={() => setShowModal(false)}
        buttons={[
          {
            textContent: "Cancel",
            className: "btn-primary",
            onClick: () => setShowModal(false),
          },
          {
            textContent: "Confirm",
            className: "btn-danger",
            onClick: handleConfirmDelete,
          },
        ]}
      />

      {!loading && error && <Navigate to="/404" />}

      {loading && (
        <div className="position-fixed top-50 left-50 w-100">
          <Spinner isLoading={loading} />
        </div>
      )}

      {!loading && advert && (
        <>
          <h2 className="text-center mb-4">Advert Detail</h2>
          <div className="container p-4">
            <div className="row ">
              <div className="col-12 col-md-4 ">
                <img
                  src={advert.photo}
                  alt={advert.name}
                  className="img-fluid rounded mx-auto d-block"
                />
              </div>
              <div className="col-12 col-md-8 border rounded d-flex flex-column justify-content-between align-items-center">
                <h3 className="text-center">{advert.name}</h3>
                <p className="d-flex align-items-center gap-2">
                  <strong>Price:</strong>
                  <span>â‚¬ {advert.price}</span>
                </p>
                <p className="d-flex align-items-center gap-2">
                  <strong className="">Category:</strong>
                  <span className="btn btn-link">{advert.tags}</span>
                </p>
                <div>
                  <Button
                    className="btn-danger"
                    onClick={() => setShowModal(true)}
                  >
                    Delete
                  </Button>
                </div>
              </div>
            </div>
          </div>
        </>
      )}
    </>
  );
};

export default AdvertPage;
