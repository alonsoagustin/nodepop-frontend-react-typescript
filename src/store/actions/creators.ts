import { NavigateFunction } from "react-router-dom";
import { Credentials, login } from "../../components/auth/service";
import {
  getTags,
  getAdvertById,
  getAllAdverts,
  deleteAdvertById,
} from "../../pages/service";
import { Advert } from "../../pages/types";
import { AppThunk } from "../store";
import Action from "./type";

export const AuthLoginPending = () => ({
  type: "AUTH_LOGIN_PENDING",
});

export const AuthLoginFulfilled = () => ({
  type: "AUTH_LOGIN_FULFILLED",
});

export const AuthLoginRejected = (error: string) => ({
  type: "AUTH_LOGIN_REJECTED",
  payload: error,
});

export const AuthLogin = (
  credentials: Credentials
): AppThunk<Promise<void>> => {
  return async function (dispatch) {
    try {
      dispatch(AuthLoginPending());
      const response = await login(credentials);
      localStorage.setItem("accessToken", response);
      dispatch(AuthLoginFulfilled());
    } catch (error) {
      const errorMessage =
        error instanceof Error ? error.message : "Ooops! Something went wrong";
      dispatch(AuthLoginRejected(errorMessage));
    }
  };
};

export const AuthLogout = (): AppThunk => (dispatch) => {
  localStorage.removeItem("accessToken");
  dispatch({ type: "AUTH_LOGOUT" });
};

export const ResetUi = (): Action => ({
  type: "RESET_UI",
});

export const AdvertsLoaded = (): AppThunk<Promise<void>> => {
  return async function (dispatch, getState) {
    const state = getState();
    if (state.adverts.loaded) return;

    dispatch(AdvertsLoadedPending());

    try {
      const data = await getAllAdverts();
      dispatch(AdvertsLoadedFulfilled(data, true));
    } catch (error) {
      const errorMessage =
        error instanceof Error ? error.message : "Ooops! Something went wrong";
      dispatch(AdvertsLoadedRejected(errorMessage));
    }
  };
};

export const AdvertsLoadedPending = () => ({
  type: "ADVERTS_LOADED_PENDING",
});

export const AdvertsLoadedFulfilled = (data: Advert[], loaded: boolean) => ({
  type: "ADVERTS_LOADED_FULFILLED",
  payload: { data, loaded },
});

export const AdvertsLoadedRejected = (error: string) => ({
  type: "ADVERTS_LOADED_REJECTED",
  payload: error,
});

export const AdvertLoaded = (advertId: string): AppThunk<Promise<void>> => {
  return async function (dispatch) {
    dispatch(AdvertsLoadedPending());
    try {
      const data = await getAdvertById(advertId);
      dispatch(AdvertsLoadedFulfilled([data], false));
    } catch (error) {
      const errorMessage =
        error instanceof Error ? error.message : "Ooops! Something went wrong";
      dispatch(AdvertsLoadedRejected(errorMessage));
    }
  };
};

export const TagsLoaded = (): AppThunk<Promise<void>> => {
  return async function (dispatch, getState) {
    const state = getState();
    if (state.tags.loaded) return;

    dispatch(TagsLoadedPending());

    try {
      const data = await getTags();
      dispatch(TagsLoadedFulfilled(data));
    } catch (error) {
      const errorMessage =
        error instanceof Error ? error.message : "Ooops! Something went wrong";
      dispatch(TagsLoadedRejected(errorMessage));
    }
  };
};

export const TagsLoadedPending = () => ({
  type: "TAGS_LOADED_PENDING",
});

export const TagsLoadedFulfilled = (data: string[]) => ({
  type: "TAGS_LOADED_FULFILLED",
  payload: data,
});

export const TagsLoadedRejected = (error: string) => ({
  type: "TAGS_LOADED_REJECTED",
  payload: error,
});

export const FilterAdvertsByTag = (tag: string) => ({
  type: "FILTER_ADVERTS_BY_TAG",
  payload: tag,
});

export const FilterAdvertsByName = (name: string) => ({
  type: "FILTER_ADVERTS_BY_NAME",
  payload: name,
});

export const AdvertCreated = (advert: Advert): Action => ({
  type: "ADVERT_CREATED",
  payload: advert,
});

export const AdvertDeleted = (
  id: string,
  navigate: NavigateFunction
): AppThunk<Promise<void>> =>
  async function (dispatch) {
    dispatch(AdvertDeletedPending());
    try {
      const advertDeleted = await deleteAdvertById(id);
      dispatch(AdvertDeletedFulfilled(advertDeleted));
      navigate("/");
    } catch (error) {
      const errorMessage =
        error instanceof Error ? error.message : "Ooops! Something went wrong";
      dispatch(AdvertDeletedRejected(errorMessage));
      navigate("/404");
    }
  };

export const AdvertDeletedPending = () => ({
  type: "ADVERT_DELETED_PENDING",
});

export const AdvertDeletedFulfilled = (advert: Advert) => ({
  type: "ADVERT_DELETED_FULFILLED",
  payload: advert,
});

export const AdvertDeletedRejected = (error: string) => ({
  type: "ADVERT_DELETED_REJECTED",
  payload: error,
});
