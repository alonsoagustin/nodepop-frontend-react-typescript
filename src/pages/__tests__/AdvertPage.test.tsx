import { fireEvent, render, screen } from "@testing-library/react";
import { Provider } from "react-redux";
import AdvertPage from "../AdvertPage";
import { BrowserRouter } from "react-router-dom";
import { AdvertDeleted } from "../../store/actions/creators";

vi.mock(
  "react-router-dom",
  async (importOriginal: () => Promise<typeof import("react-router-dom")>) => {
    const actual = await importOriginal(); // Importamos la implementación original de react-router-dom
    return {
      ...actual, // Mantener la implementación original de los elementos
      useParams: vi.fn().mockReturnValue({ advertId: "mockId" }), // Mockeamos solo useParams
    };
  }
);

vi.mock("../../store/actions/creators", () => ({
  AdvertDeleted: vi.fn(),
}));

const advert = {
  id: "mockId",
  createdAt: "mockDate",
  name: "mockName",
  sale: true,
  price: 111,
  tags: ["mockTag"],
  photo: "http://mockPhoto.jpg",
};

const state = {
  auth: true,
  adverts: { data: [advert], loaded: false },
  ui: { error: "", loading: false },
};

const renderComponent = () =>
  render(
    <BrowserRouter>
      <Provider
        store={{
          getState: () => state,
          subscribe: vi.fn(),
          dispatch: vi.fn(),
          replaceReducer: vi.fn(),
          [Symbol.observable]: vi.fn(),
        }}
      >
        <AdvertPage />
      </Provider>
    </BrowserRouter>
  );

describe("AdvertPage", () => {
  test("should render advert page", () => {
    const { container } = renderComponent();
    expect(container).toMatchSnapshot();
  });

  test("should open modal and dispatch delete action", () => {
    renderComponent();
    const deleteButton = screen.getByRole("button");

    expect(deleteButton).toBeInTheDocument();
    expect(deleteButton).toHaveTextContent(/Delete/);

    fireEvent.click(deleteButton);

    const modal = screen.getByRole("dialog");

    expect(modal).toBeInTheDocument();

    const confirmButton = screen.getByText(/Confirm/);

    expect(confirmButton).toBeInTheDocument();

    fireEvent.click(confirmButton);

    expect(AdvertDeleted).toHaveBeenCalled();
    expect(AdvertDeleted).toHaveBeenCalledWith("mockId", expect.any(Function));
  });
});
