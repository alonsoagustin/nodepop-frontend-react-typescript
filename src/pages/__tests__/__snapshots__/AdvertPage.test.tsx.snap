// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AdvertPage > should render advert page 1`] = `
<div>
  <h2
    class="text-center mb-4"
  >
    Advert Detail
  </h2>
  <div
    class="container p-4"
  >
    <div
      class="row "
    >
      <div
        class="col-12 col-md-4 "
      >
        <img
          alt="mockName"
          class="img-fluid rounded mx-auto d-block"
          src="http://mockPhoto.jpg"
        />
      </div>
      <div
        class="col-12 col-md-8 border rounded d-flex flex-column justify-content-between align-items-center"
      >
        <h3
          class="text-center"
        >
          mockName
        </h3>
        <p
          class="d-flex align-items-center gap-2"
        >
          <strong>
            Price:
          </strong>
          <span>
            â‚¬ 
            111
          </span>
        </p>
        <p
          class="d-flex align-items-center gap-2"
        >
          <strong
            class=""
          >
            Category:
          </strong>
          <span
            class="btn btn-link"
          >
            mockTag
          </span>
        </p>
        <div>
          <button
            class="btn btn-danger"
          >
            Delete
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
